From 0c9d642b7ad0a90e4863829151ede21974abc102 Mon Sep 17 00:00:00 2001
From: Christophe Dumez <christophe.dumez@intel.com>
Date: Mon, 20 Jun 2011 17:59:58 +0300
Subject: [PATCH] MemoryCalendar::rawEventsForDate() Fix event duplication
 issue

Recurring events that start on the given date are returned twice
because the first loop fails to ignore recurring events.

MemoryCalendar::rawTodosForDate() Fix todo duplication issue

Recurring todos that start on the given date are returned twice
because the first loop fails to ignore recurring todos.
---
 src/memorycalendar.cpp |   22 +++++++++++++---------
 1 files changed, 13 insertions(+), 9 deletions(-)

diff --git a/src/memorycalendar.cpp b/src/memorycalendar.cpp
index 285a2e4..01e98c6 100644
--- a/src/memorycalendar.cpp
+++ b/src/memorycalendar.cpp
@@ -384,7 +384,8 @@ Todo::List MemoryCalendar::rawTodosForDate( const QDate &date ) const
     d->mIncidencesForDate[Incidence::TypeTodo].constFind( dateStr );
   while ( it != d->mIncidencesForDate[Incidence::TypeTodo].constEnd() && it.key() == dateStr ) {
     t = it.value().staticCast<Todo>();
-    todoList.append( t );
+    if ( !t->recurs() )
+      todoList.append( t );
     ++it;
   }
 
@@ -555,14 +556,17 @@ Event::List MemoryCalendar::rawEventsForDate( const QDate &date,
   KDateTime kdt( date, ts );
   while ( it != d->mIncidencesForDate[Incidence::TypeEvent].constEnd() && it.key() == dateStr ) {
     ev = it.value().staticCast<Event>();
-    KDateTime end( ev->dtEnd().toTimeSpec( ev->dtStart() ) );
-    if ( ev->allDay() ) {
-      end.setDateOnly( true );
-    } else {
-      end = end.addSecs( -1 );
-    }
-    if ( end >= kdt ) {
-      eventList.append( ev );
+    // Skip recurring events
+    if ( !ev->recurs() ) {
+      KDateTime end( ev->dtEnd().toTimeSpec( ev->dtStart() ) );
+      if ( ev->allDay() ) {
+        end.setDateOnly( true );
+      } else {
+        end = end.addSecs( -1 );
+      }
+      if ( end >= kdt ) {
+        eventList.append( ev );
+      }
     }
     ++it;
   }
-- 
1.7.4.4

